src/main/cc/wfa/virtual_people/core/model/utils/consistent_hash.cc

```diff
@@ -0,0 +1,35 @@
+// Copyright 2021 The Cross-Media Measurement Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "wfa/virtual_people/core/model/utils/consistent_hash.h"
+
+#include <cstdint>
+
+namespace wfa_virtual_people {
+
+// This implementation is a copy and formatting of Figure 1 on page 2 in the
+// published paper:
+//   https://arxiv.org/pdf/1406.2294.pdf
+int32_t JumpConsistentHash(uint64_t key, int32_t num_buckets) {
+  int64_t b = -1, j = 0;
+  while (j < num_buckets) {
+    b = j;
+    key = key * 2862933555777941757ULL + 1;
+    j = (b + 1) * (static_cast<double>(1LL << 31) /
```

Comment:
i know this is from the code in the paper. from c++ perspective, we only need  `uint64` and `double` for the `1LL<<31 / key>>33 ...` only, how about making some slight change as the following:

```
int32_t b = -1;
for (int32_t j = 0; j < num_bucktes; ) {
  b = j;
  ...
  j = static_cast<int32_t>( ... );
}
```

Reply:
Though this is very unlikely to happen, if num_bucktes is std::numeric_limits<int>::max, j < num_buckets will never be fulfilled if j is int32_t.

Reply:
i meant, just to keep the type consistent in the code.   sure, we don't need to worry about the overflow

Reply:
I think int32_t does not work for this algorithm. Changed to cast the return to int32_t explicitly.

Reply:
don't change the return type here. see my code snippet above, `j = static_cast<int32_t>(...)` is the key part

Reply:
I mean we cannot cast j to int32_t. If we cast j to int32_t, and the num_buckets is very big, the condition "j >= num_bucktes" is hard to fulfill, and the while/for loop will last too long. Furthermore, if num_buckets = INT_MAX, the while/for loop will never end.

Reply:
what's the difference between int32 and int64 for `j` here?   anyway, j needs to reach `num_buckets` in jumps

Reply:
"j needs to reach num_buckets in jumps", this is what I mean. j should be able to reach larger than maximum possible value of num_buckets.

Reply:
the code can be like this:
```
// Generates the bucket index in range of [0, numBuckets).                                                                              
int32_t JumpConsistentHash(uint64_t key, int32_t num_buckets) {
  int32_t b = -1;
  for (int32_t j = 0; j < num_buckets; ) {
    b = j;
    key = key * 2862933555777941757ULL + 1;
    j = (j+1) * (static_cast<double>(1LL << 31) /
                 static_cast<double>((key >> 33) + 1));
  }
  return b;
}
```

Reply:
How about
```
int32_t JumpConsistentHash(uint64_t key, int32_t num_buckets) {
  int32_t b = -1;
  for (int64_t j = 0; j < num_buckets; ) {
    b = j;
    key = key * 2862933555777941757ULL + 1;
    j = (j+1) * (static_cast<double>(1LL << 31) /
                 static_cast<double>((key >> 33) + 1));
  }
  return b;
}
```
Though j will reach num_buckets finally, the function is very slow when num_buckets is big. We will call this function for every event, so the performance will have great impact.

Reply:
Just changed b from int64_t to int32_t, and still keep j as int64_t due to performance consideration.

Reply:
`int64_t` doesn't look that rigorous.  how about just adding "j>=0" to prevent overflow?

```
// Generates the bucket index in range of [0, numBuckets).                                                                              
int32_t JumpConsistentHash(uint64_t key, int32_t num_buckets) {
  int32_t b = -1;
  for (int32_t j = 0; j < num_buckets && j >= 0; ) {
    // j>=0 to prevent overflow
    b = j;
    key = key * 2862933555777941757ULL + 1;
    j = (j+1) * (static_cast<double>(1LL << 31) /
                 static_cast<double>((key >> 33) + 1));
  }
  return b;
}
```

Reply:
In the code
```
    j = (j+1) * (static_cast<double>(1LL << 31) /
                 static_cast<double>((key >> 33) + 1));
```
The value of right side is a double floating number. In c++, the conversion from floating number to integer with overflow is unidentified. So it is dangerous to assume that the overflow will cause j to be negative.

I think the behavior of setting j as int64_t was verified, as it is used in the code from the published paper. Changing the type to int32_t will cause the return to be different.

Reply:
Makes sense! LGTM! Add a comment “int64 j to handle int32 overflow”

Reply:
Comment added. Thanks!